// appium-espresso-driver/build/test/functional/helpers/session.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _wd = require('wd');

var _wd2 = _interopRequireDefault(_wd);

var _ = require('../../..');

var HOST = '0.0.0.0',
    PORT = 4994;
var MOCHA_TIMEOUT = 60 * 1000 * (process.env.TRAVIS ? 10 : 4);

var driver = undefined,
    server = undefined;

function initDriver() {
  return _regeneratorRuntime.async(function initDriver$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        driver = _wd2['default'].promiseChainRemote(HOST, PORT);
        context$1$0.next = 3;
        return _regeneratorRuntime.awrap((0, _.startServer)(PORT, HOST));

      case 3:
        server = context$1$0.sent;
        return context$1$0.abrupt('return', driver);

      case 5:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function initSession(caps) {
  var serverRes;
  return _regeneratorRuntime.async(function initSession$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(initDriver());

      case 2:
        context$1$0.next = 4;
        return _regeneratorRuntime.awrap(driver.init(caps));

      case 4:
        serverRes = context$1$0.sent;

        if (!caps.udid && !caps.fullReset && serverRes[1].udid) {
          caps.udid = serverRes[1].udid;
        }

        // await driver.setImplicitWaitTimeout(5000);

        return context$1$0.abrupt('return', driver);

      case 7:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function deleteSession() {
  return _regeneratorRuntime.async(function deleteSession$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.prev = 0;
        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(driver.quit());

      case 3:
        context$1$0.next = 7;
        break;

      case 5:
        context$1$0.prev = 5;
        context$1$0.t0 = context$1$0['catch'](0);

      case 7:
        context$1$0.prev = 7;
        context$1$0.next = 10;
        return _regeneratorRuntime.awrap(server.close());

      case 10:
        context$1$0.next = 14;
        break;

      case 12:
        context$1$0.prev = 12;
        context$1$0.t1 = context$1$0['catch'](7);

      case 14:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[0, 5], [7, 12]]);
}

exports.initDriver = initDriver;
exports.initSession = initSession;
exports.deleteSession = deleteSession;
exports.HOST = HOST;
exports.PORT = PORT;
exports.MOCHA_TIMEOUT = MOCHA_TIMEOUT;
